### ISSUE TITLE
Can't get persistent sessions to work

### ISSUE THREAD

**Issue by kidbombay** [danschultzer/pow#202]
I'm using Pow with the persistent sessions extension, but I don't see the cookie saved

Config
```
config :my_app, :pow,
  user: MyApp.Users.User,
  repo: MyApp.Repo,
  web_module: MyAppWeb,
  routes_backend: MyAppWeb.Pow.Routes,
  cache_store_backend: MyAppWeb.Pow.PowRedisCache,
  mailer_backend: MyAppWeb.Pow.PowMailer,
  extensions: [PowResetPassword, PowPersistentSession]
```

Endpoint
```
plug Pow.Plug.Session, otp_app: :my_app
plug PowPersistentSession.Plug.Cookie
```
User
```
defmodule MyApp.Users.User do
  use Ecto.Schema
  use Pow.Ecto.Schema
  use Pow.Extension.Ecto.Schema,
    extensions: [PowResetPassword, PowPersistentSession]
```

Any ideas?

**Comment by danschultzer** [comment]
It looks right, and by default it should set the persistent session unless [you've a "Remember me" checkbox](https://github.com/danschultzer/pow/tree/master/lib/extensions/persistent_session#configuration). I'll do a manual test here just to make sure it should be working as expected.

**Comment by kidbombay** [comment]
I added the checkbox as well and even when true I don't see the cookie being saved.

**Comment by danschultzer** [comment]
If the checkbox is removed does it work then?

**Comment by kidbombay** [comment]
No it doesn't work in either case.

**Comment by danschultzer** [comment]
You're missing the controller callbacks:

```elixir
config :my_app, :pow,
  user: MyApp.Users.User,
  repo: MyApp.Repo,
  web_module: MyAppWeb,
  routes_backend: MyAppWeb.Pow.Routes,
  cache_store_backend: MyAppWeb.Pow.PowRedisCache,
  mailer_backend: MyAppWeb.Pow.PowMailer,
  extensions: [PowResetPassword, PowPersistentSession],
  controller_callbacks: Pow.Extension.Phoenix.ControllerCallbacks
```

The controller callbacks will interject the persistent session logic in the session create action 😄

**Comment by danschultzer** [comment]
There's some more info in the readme: https://github.com/danschultzer/pow#add-extensions-support

**Comment by lud** [comment]
Hi, just a quick question, I'm not sure I should open a new issue :

I added the persistent session extension and the controllers callbacks.
I can see in this issue that I should do something like that :

```
  use Pow.Extension.Ecto.Schema,
    extensions: [PowPersistentSession]
```

But it seems not advised in your README so I think it's for a previous version, so I dit **not** add it and it seems fine.

I would like sessions to persist over app restarts because when developping I have to stop multiple GenServers so it's simpler to just restart the VM at the moment.

Sessions doesn't seem to persist over app restarts. So I changed the `cache_store_backend` for a persisted ETS table, and it works.

My question is : am I right ? does the persistent sessions use this ETS cache so it is normal to have to put a custom cache ?

Thank you.

**Comment by danschultzer** [comment]
Yeah, it's unrelated to persistent session. With the default EtsCache, all cache store data only exists in memory, so changing to a persisted ETS table is the right solution. However, you can also use the MnesiaCache that's recommended for production as per the readme: https://github.com/danschultzer/pow#cache-store

For apps running in read-only filesystems, redis can be used: https://github.com/danschultzer/pow/blob/master/guides/redis_cache_store_backend.md

As for PowPersistentSession in the user schema module, it's not needed since there is no ecto schema modules in PowPersistentSession, but I usually recommend adding it just to keep your configuration consistent with the readme instructions: https://github.com/danschultzer/pow/blob/master/README.md#add-extensions-support

**Comment by lud** [comment]
This is just a quick fix to be able to restart often ; I do not want to handle mnesia right now but yes, that would be the next step.

But actually I think that I am confusing two things. The persistent session would be to store session data, whereas I just need to keep a user logged in on restarts, which could be possible with a mere cookie. Is there a way to keep users logged in over restarts and let the session be created automatically ?

Thank you for your time :)

**Comment by danschultzer** [comment]
Yeah, you won't need persistent session for that. To be clear, persistent session is only to make it possible for users to remain logged in for a long time (e.g. they'll remained logged in when they restart their computer, or visit the website a few days later).

To ensure that sessions remain active when restarting the server you just need to have a persistent cache backend where the session data is stored. The sessions are short-lived, and will expire after 30 minutes of inactivity.

**Comment by lud** [comment]
Ok but the automatic log-in is not baked in Pow, right ?
As I store nothing in sessions the persistent cache should remain lightweight, so I think I'll keep it that way for now.

Thanks again !

**Comment by danschultzer** [comment]
If you enable PowPersistentSession extension, then it'll automatically log in. But in Pow core, no, it'll only have the short-lived sessions 😄

**Comment by lud** [comment]
Right on the same page, thank you.

**Comment by lud** [comment]
Hey, 

So I am back to this project after a long time and my persistent storage does not work. I wonder why Pow does not provide "remember me" functionality withour relying on persisten sessions out of the box.

I mean an auto log-in that automatically creates a new session if you checked the "remember me" box, and *not* a persistent session where your session data would be persisted server side.

It should be simple enough to implement, and I can have a look at it but I guess I'd need some directions.

**Comment by danschultzer** [comment]
> I mean an auto log-in that automatically creates a new session if you checked the "remember me" box, and _not_ a persistent session where your session data would be persisted server side.

That's what [PowPersistentSession does](https://hexdocs.pm/pow/1.0.14/pow_persistent_session.html#content). It only stores the token and the user id and will load the user into a session when used, invalidating the token in the process. You can switch it out with an encrypted JWT if you wish to not store anything server-side, but it isn't a good idea since it'll be possible to use that token numerous times until TTL reached.

I assume the issue you experience is that the cache date doesn't persist between server restarts. The [production checklist](https://hexdocs.pm/pow/1.0.14/production_checklist.html#content) guide goes into this, and using the MnesiaCache would be the easiest way to resolve that.

**Comment by lud** [comment]
Hey, thank you.

Yes I set mnesia up, it seems to work without having to create a schema.

I was confused because using a cookie (as in `plug PowPersistentSession.Plug.Cookie`) let me think that there is no cache involved. Back in the days I used to generate a random token (arbitraty binary, not a JWT) and store that into a field in the user table, so you could automatically re-login users and recreate a session. You can obviously generate a new token at that time if you wanted the token to be usable only once. Also we used a dedicated table for those tokens to allow re-login from multiple devices.

Why did you chose against this method ?

**Comment by danschultzer** [comment]
> I was confused because using a cookie (as in `plug PowPersistentSession.Plug.Cookie`) let me think that there is no cache involved.

Oh I see. Yeah, this is the same as `Pow.Plug.Session` storing a cookie with the session data used to pull the credentials from the cache.

> Why did you chose against this method ?

I would either store all session data in the DB or none.

You may have multiple devices that you are logged in with, so you would need to be able to set multiple persistent session tokens for the user, and you would also have to deal with TTL and ability to flush tokens at request.

Using the same cache engine as the session credentials also makes it easy to add metadata to the token, such as session fingerprint to match up with the session that created the persistent session token. This makes it easy to flush any active sessions when the persistent session token is used.

No issue storing it in the DB, but a single column for the user wouldn't be sufficiently secure or flexible.

**Comment by lud** [comment]
Interesting, thank you :)

**Comment by lud** [comment]
Hey, one quick question after reading this thread again. Do you confirm that the mnesia cache stores only limited amount of data (user id and password hash) so it is lightweight ?

I start `Pow.Store.Backend.MnesiaCache` in my supervision tree but I did not create any schema, so I guess everything is in memory.
We use docker containers so I still need to figure out how to use mnesia with disk copies on docker volumes.

**Comment by danschultzer** [comment]
> Do you confirm that the mnesia cache stores only limited amount of data (user id and password hash) so it is lightweight ?

It stores the whole struct, but only once per user and the records have a TTL of 30 minutes. The schema is both in memory and written to disk, so if you restart the server it won't lose existing sessions.

If you run multiple containers then you need to set it up as a cluster: https://github.com/danschultzer/pow#cache-store

More here in the docs: https://hexdocs.pm/pow/1.0.16/Pow.Store.Backend.MnesiaCache.html#module-distribution

**Comment by lud** [comment]
Thank you :)

**Comment by lud** [comment]
But how comes we do not have to `mnesia:create_schema(…)` when using mnesia with disk copies ?

**Comment by danschultzer** [comment]
I use [` :mnesia.change_table_copy_type/3`](https://github.com/danschultzer/pow/blob/v1.0.16/lib/pow/store/backend/mnesia_cache.ex#L354) which does the same.

**Comment by lud** [comment]
I see. The mnesia docs tell us that the mnesia application must not be started when we create a schema. But using this function seems to be ok if mnesia is started. Whatever :)
